### PR TITLE
Moi v0.10

### LINKED ISSUE #172: Support looking up variable/constraints by name
Important to have access to more MOI tests and for the user experience.

### PR DESCRIPTION
**Pull request by Wikunia**
Currently there are still two problems:
- I'm unsure how to use `MOIT.runtests` (maybe without the bride optimizer?)
- I have a weird `LowerBoundAlreadySet` error in one feasibility pump problem. I'm not too familiar with the code by @blegat from removing the JuMP dependency on that. Will come back to it next Friday probably.

Currently testing involves:
```
] add JuMP#master  
add Cbc#master
add Ipopt#master
```
Not sure how to do compat only for test dependencies. 

Closes #172

### CODE REVIEW

**Comment by ccoffrin** [comment]
@odow any tips on the `MOIT.runtests` point?

**Comment by blegat** [comment]
> maybe without the bride optimizer?

Juniper supports everything so I don't think that the bridges would be used for:
https://github.com/lanl-ansi/Juniper.jl/blob/3a30ee5298d0dae426348da840c994eda901b985/src/MOI_wrapper/MOI_wrapper.jl#L358

**Comment by Wikunia** [comment]
Quite a lot of tests seem to be changed and break currently like getting a variable index by name. Kind of surprised that this should be handled by the solver itself. 
Also I'm not a fan of returning the SolverVersion as I currently do because at one point in the future it will be definitely out of sync. Maybe there is a way to obtain the current version number though hopefully without an extra package dependency.

**Comment by blegat** [comment]
For names, you could do
```julia
MOI.Test.Config(
            exclude = Any[
                MOI.ConstraintName,
                MOI.VariableName,
                MOI.Name,
            ],
        )
```

File: src/MOI_wrapper/MOI_wrapper.jl [hunk 1]
```diff
@@ -15,11 +15,12 @@ end
 MOI.is_valid(model::Optimizer, index::MOI.Index) = MOI.is_valid(model.model_cache, index)
 
 MOI.get(::Optimizer, ::MOI.SolverName) = "Juniper"
+MOI.get(::Optimizer, ::MOI.SolverVersion) = "v0.7.0"
```
Comment (odow) [inline src/MOI_wrapper/MOI_wrapper.jl:18]: You can always not implement it and exclude the test.

File: test/MOI_wrapper.jl [hunk 2]
```diff
@@ -49,5 +53,5 @@ const config = MOIT.TestConfig(atol=1e-4, rtol=1e-4, optimal_status=MOI.LOCALLY_
                 "solve_result_index", # no support for `MOI.ConstraintPrimal` atm
                 "update_dimension_nonnegative_variables", # currently no support for ConstraintFunction 
                ]
-    MOIT.unittest(bridged, config, exclude)
+    MOIT.runtests(optimizer, config; exclude=exclude)
```
Comment (odow) [inline test/MOI_wrapper.jl:56]: Run the tests in a caching optimizer, that should fix a lot of the `loadfromstring!` errors.

```Julia
MOI.Utilities.CachingOptimizer(
    MOI.Utilities.UniversalFallback(MOI.Utilities.Model{Float64}()),
    optimizer,
)
```

**Comment by odow** [comment]
> Quite a lot of tests seem to be changed and break currently

To the contrary, because tests were opt-in, most solvers did not test against the full API of MathOptInterface, and missed implementing some critical features. The new approach is opt-out, which has exposed a lot of issues in the wrappers (many more than I expected!).

**Comment by Wikunia** [comment]
> For names, you could do
> 
> ```julia
> MOI.Test.Config(
>     exclude = Any[
>         MOI.ConstraintName,
>         MOI.VariableName,
>         MOI.Name,
>     ],
> )
> ```

`ConstraintName` and `VariableName` is already excluded in there. 
I've also tried to exclude `ConstraintPrimal` there but still getting errors regarding that.

**Comment by odow** [comment]
Just wrap it in a cachingoptimizer. It looks like not all the tests bail early if constraint name isn't supported.

**Comment by Wikunia** [comment]
that doesn't seem to work or am I doing something wrong?

**Comment by odow** [comment]
I'll take a look.

**Comment by odow** [comment]
Now there is a single failure:
```
==================================
446
FP: Integer Test 2
447
==================================
448
FP: Integer test2: Error During Test at /home/runner/work/Juniper.jl/Juniper.jl/test/fpump.jl:123
449
  Got exception outside of a @test
450
  MathOptInterface.LowerBoundAlreadySet{MathOptInterface.EqualTo{Float64},MathOptInterface.EqualTo{Float64}}: Cannot add `VariableIndex`-in-`MathOptInterface.EqualTo{Float64}` constraint for variable MathOptInterface.VariableIndex(4) as a `VariableIndex`-in-`MathOptInterface.EqualTo{Float64}` constraint was already set for this variable and both constraints set a lower bound.
451
  Stacktrace:
452
   [1] add_constraint at /home/runner/.julia/packages/Ipopt/QF8Lc/src/MOI_wrapper.jl:812 [inlined]
453
   [2] add_constraint(::MathOptInterface.Bridges.LazyBridgeOptimizer{Ipopt.Optimizer}, ::MathOptInterface.VariableIndex, ::MathOptInterface.EqualTo{Float64}) at /home/runner/.julia/packages/MathOptInterface/jPhq9/src/Bridges/bridge_optimizer.jl:1546
454
   [3] _copy_constraints(::MathOptInterface.Bridges.LazyBridgeOptimizer{Ipopt.Optimizer}, ::Juniper.FixVariables{Float64,Juniper.IntegerRelaxation{Juniper.Optimizer}}, ::MathOptInterface.Utilities.IndexMap, ::Array{MathOptInterface.ConstraintIndex{MathOptInterface.VariableIndex,MathOptInterface.EqualTo{Float64}},1}) at /home/runner/.julia/packages/MathOptInterface/jPhq9/src/Utilities/copy.jl:241
455
   [4] _pass_constraints(::MathOptInterface.Bridges.LazyBridgeOptimizer{Ipopt.Optimizer}, ::Juniper.FixVariables{Float64,Juniper.IntegerRelaxation{Juniper.Optimizer}}, ::MathOptInterface.Utilities.IndexMap, ::Array{Any,1}) at /home/runner/.julia/packages/MathOptInterface/jPhq9/src/Utilities/copy.jl:296
456
   [5] default_copy_to(::MathOptInterface.Bridges.LazyBridgeOptimizer{Ipopt.Optimizer}, ::Juniper.FixVariables{Float64,Juniper.IntegerRelaxation{Juniper.Optimizer}}) at /home/runner/.julia/packages/MathOptInterface/jPhq9/src/Utilities/copy.jl:435
457
   [6] #copy_to#7 at /home/runner/.julia/packages/MathOptInterface/jPhq9/src/Bridges/bridge_optimizer.jl:421 [inlined]
458
   [7] copy_to(::MathOptInterface.Bridges.LazyBridgeOptimizer{Ipopt.Optimizer}, ::Juniper.FixVariables{Float64,Juniper.IntegerRelaxation{Juniper.Optimizer}}) at /home/runner/.julia/packages/MathOptInterface/jPhq9/src/Bridges/bridge_optimizer.jl:421
459
   [8] generate_real_nlp(::Juniper.Optimizer, ::Juniper.JuniperProblem, ::Array{Float64,1}; random_start::Bool) at /home/runner/work/Juniper.jl/Juniper.jl/src/fpump.jl:159
460
   [9] generate_real_nlp(::Juniper.Optimizer, ::Juniper.JuniperProblem, ::Array{Float64,1}) at /home/runner/work/Juniper.jl/Juniper.jl/src/fpump.jl:152
461
   [10] fpump(::Juniper.Optimizer, ::Juniper.JuniperProblem) at /home/runner/work/Juniper.jl/Juniper.jl/src/fpump.jl:351
462
   [11] optimize!(::Juniper.Optimizer) at /home/runner/work/Juniper.jl/Juniper.jl/src/MOI_wrapper/MOI_wrapper.jl:285
463
   [12] optimize! at /home/runner/.julia/packages/MathOptInterface/jPhq9/src/MathOptInterface.jl:81 [inlined]
464
   [13] optimize!(::MathOptInterface.Utilities.CachingOptimizer{Juniper.Optimizer,MathOptInterface.Utilities.UniversalFallback{MathOptInterface.Utilities.GenericModel{Float64,MathOptInterface.Utilities.ObjectiveContainer{Float64},MathOptInterface.Utilities.VariablesContainer{Float64},MathOptInterface.Utilities.ModelFunctionConstraints{Float64}}}}) at /home/runner/.julia/packages/MathOptInterface/jPhq9/src/Utilities/cachingoptimizer.jl:285
465
   [14] optimize!(::MathOptInterface.Bridges.LazyBridgeOptimizer{MathOptInterface.Utilities.CachingOptimizer{Juniper.Optimizer,MathOptInterface.Utilities.UniversalFallback{MathOptInterface.Utilities.GenericModel{Float64,MathOptInterface.Utilities.ObjectiveContainer{Float64},MathOptInterface.Utilities.VariablesContainer{Float64},MathOptInterface.Utilities.ModelFunctionConstraints{Float64}}}}}) at /home/runner/.julia/packages/MathOptInterface/jPhq9/src/Bridges/bridge_optimizer.jl:348
466
   [15] optimize! at /home/runner/.julia/packages/MathOptInterface/jPhq9/src/MathOptInterface.jl:81 [inlined]
467
   [16] optimize!(::MathOptInterface.Utilities.CachingOptimizer{MathOptInterface.Bridges.LazyBridgeOptimizer{MathOptInterface.Utilities.CachingOptimizer{Juniper.Optimizer,MathOptInterface.Utilities.UniversalFallback{MathOptInterface.Utilities.GenericModel{Float64,MathOptInterface.Utilities.ObjectiveContainer{Float64},MathOptInterface.Utilities.VariablesContainer{Float64},MathOptInterface.Utilities.ModelFunctionConstraints{Float64}}}}},MathOptInterface.Utilities.UniversalFallback{MathOptInterface.Utilities.GenericModel{Float64,MathOptInterface.Utilities.ObjectiveContainer{Float64},MathOptInterface.Utilities.VariablesContainer{Float64},MathOptInterface.Utilities.ModelFunctionConstraints{Float64}}}}) at /home/runner/.julia/packages/MathOptInterface/jPhq9/src/Utilities/cachingoptimizer.jl:285
468
   [17] optimize!(::Model, ::Nothing; bridge_constraints::Bool, ignore_optimize_hook::Bool, kwargs::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}) at /home/runner/.julia/packages/JuMP/yP8fy/src/optimizer_interface.jl:195
469
   [18] optimize! at /home/runner/.julia/packages/JuMP/yP8fy/src/optimizer_interface.jl:167 [inlined] (repeats 2 times)
470
   [19] solve(::Model) at /home/runner/work/Juniper.jl/Juniper.jl/test/runtests.jl:52
471
   [20] top-level scope at /home/runner/work/Juniper.jl/Juniper.jl/test/fpump.jl:149
472
   [21] top-level scope at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.5/Test/src/Test.jl:1119
473
   [22] top-level scope at /home/runner/work/Juniper.jl/Juniper.jl/test/fpump.jl:124
474
   [23] top-level scope at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.5/Test/src/Test.jl:1119
475
   [24] top-level scope at /home/runner/work/Juniper.jl/Juniper.jl/test/fpump.jl:8
476
   [25] include(::String) at ./client.jl:457
477
   [26] top-level scope at /home/runner/work/Juniper.jl/Juniper.jl/test/runtests.jl:113
478
   [27] top-level scope at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.5/Test/src/Test.jl:1119
479
   [28] top-level scope at /home/runner/work/Juniper.jl/Juniper.jl/test/runtests.jl:108
480
   [29] include(::String) at ./client.jl:457
481
   [30] top-level scope at none:6
482
   [31] eval(::Module, ::Any) at ./boot.jl:347
483
   [32] exec_options(::Base.JLOptions) at ./client.jl:272
484
   [33] _start() at ./client.jl:506
```

**Comment by odow** [comment]
@Wikunia have the feasibility pump tests always been flakey? 

Linux:

image

Mac and Windows:

image

**Comment by Wikunia** [comment]
Thanks a lot @odow 
I'm not sure about this particular test but some tests are unfortunately time dependent and then it can be the case that for whatever reason thae one system is slower and doesn't find the incumbent in time. I'm sure there are better ways to handle those things.

**Comment by odow** [comment]
Okay, fair enough. If/when I start working on nonlinear stuff I can take a longer look at some maintenance of Juniper. I've opened some issues to remind myself in the future.

This is good to go by me for the near term. Things are working and updated, at least.

**Comment by ccoffrin** [comment]
Looks like tests are passing!  Amazing, thanks @odow.

@Wikunia, would you like to merge and tag?  I could also lend a hand with this if needed.

**Pull request merged**